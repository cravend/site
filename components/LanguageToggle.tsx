import { Box } from "theme-ui";
import React, { FC } from "react";
import useLocalization from "../src/hooks/useLocalization";
import Link from "./Link";

const LanguageToggle: FC = () => {
  const [t, , otherLocale] = useLocalization();

  return (
    <Box
      p={[0, 3]}
      sx={{
        position: ["relative", "absolute"],
        top: 0,
        right: 0,
      }}
    >
      <Link to="/" locale={otherLocale} color={["primary", "white"]}>
        {t("toggleLocale")}
      </Link>
    </Box>
  );
};

LanguageToggle.defaultProps = {
  color: undefined,
};

export default LanguageToggle;
