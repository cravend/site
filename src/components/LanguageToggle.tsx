import useLocalization from "../i18n/useLocalization";
import styles from "../styles/modules/LanguageToggle.module.scss";
import Link from "./Link";

const LanguageToggle = () => {
  const [t, , otherLocale] = useLocalization();

  return (
    <div className={styles.container}>
      <Link to="/" locale={otherLocale}>
        {t("toggle_locale")}
      </Link>
    </div>
  );
};

export default LanguageToggle;
