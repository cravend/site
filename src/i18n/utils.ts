import { locales, defaultLocale, RTL_LOCALES } from "./config";

import type { Locale } from "./config";
import type { NextRouter } from "next/router";

/**
 * @description This function is used to check if a string is a valid locale.
 *
 * @param localeToTest The string to check.
 * @returns True if the string is a valid locale, false otherwise.
 */
export const isLocale = (localeToTest?: string): localeToTest is Locale =>
  locales.some((locale) => locale === localeToTest);

/**
 * @description This function is used to get the current locale.
 *
 * If a custom locale is set by \@next/router, it will be returned.
 * If no custom locale is set, the router's default locale will be returned.
 * If the router's default locale isn't set, the default locale will be returned.
 *
 * @param router The NextRouter instance.
 * @returns The current locale.
 */
export const getLocale = (router: NextRouter | null) => {
  if (!router) return defaultLocale;
  if (isLocale(router.locale)) return router.locale;
  if (isLocale(router.defaultLocale)) return router.defaultLocale;
  return defaultLocale;
};

/**
 * @description This function is used to toggle the current locale.
 *
 * @param locale The current locale to be toggled.
 * @returns The new locale.
 */
export const toggleLocale = (locale: Locale) =>
  locale === locales[0] ? locales[1] : locales[0];

/**
 * @description This function is used to get the text direction of the current locale.
 *
 * @param locale The locale to be checked
 * @returns True if the locale is read right-to-left, false otherwise
 */
export const isRtl = (locale: Locale) => RTL_LOCALES.includes(locale);
