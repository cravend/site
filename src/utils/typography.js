import Typography from 'typography'

const typography = new Typography({
  baseLineHeight: 1.666,
  headerFontFamily: ['Poppins', 'Helvetica', 'Arial', 'sans-serif'],
  bodyFontFamily: ['Poppins', 'Helvetica', 'Arial', 'sans-serif'],
  googleFonts: [
    {
      name: 'Poppins',
      styles: ['300', '600', '700']
    }
  ]
})

export default typography
