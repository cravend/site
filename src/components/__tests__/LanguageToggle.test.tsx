import { screen, render, fireEvent } from "../../tests/utils";
import LanguageToggle from "../LanguageToggle";

describe('<LanguageToggle /> { locale: "en" }', () => {
  it("renders", () => {
    expect.assertions(1);
    const { container } = render(<LanguageToggle />, {
      router: { locale: "en" },
    });
    expect(container).toMatchSnapshot();
  });

  it("displays a link to read in French", () => {
    expect.assertions(1);
    render(<LanguageToggle />, { router: { locale: "en" } });
    expect(screen.getByRole("link", { name: "Français" })).toBeInTheDocument();
  });

  it("routes to French", () => {
    expect.assertions(2);
    const pushMock = jest.fn();
    render(<LanguageToggle />, { router: { locale: "en", push: pushMock } });

    fireEvent.click(screen.getByRole("link", { name: "Français" }));
    expect(pushMock).toHaveBeenCalledTimes(1);
    expect(pushMock).toHaveBeenLastCalledWith("/", "/", { locale: "fr" });
  });
});

describe('<LanguageToggle /> { locale: "fr" }', () => {
  it("renders", () => {
    expect.assertions(1);
    const { container } = render(<LanguageToggle />, {
      router: { locale: "fr" },
    });
    expect(container).toMatchSnapshot();
  });

  it("displays a link to read in English", () => {
    expect.assertions(1);
    render(<LanguageToggle />, { router: { locale: "fr" } });
    expect(screen.getByRole("link", { name: "English" })).toBeInTheDocument();
  });

  it("routes to English", () => {
    expect.assertions(2);
    const pushMock = jest.fn();
    render(<LanguageToggle />, { router: { locale: "fr", push: pushMock } });

    fireEvent.click(screen.getByRole("link", { name: "English" }));
    expect(pushMock).toHaveBeenCalledTimes(1);
    expect(pushMock).toHaveBeenLastCalledWith("/", "/", { locale: "en" });
  });
});
