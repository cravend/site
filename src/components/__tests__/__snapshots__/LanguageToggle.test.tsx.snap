// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LanguageToggle /> { locale: "en" } renders 1`] = `
<div>
  <div
    class="css-hckpid"
  >
    <a
      class="css-zdzv7c"
      href="/"
    >
      Lire en fran√ßais
    </a>
  </div>
</div>
`;

exports[`<LanguageToggle /> { locale: "fr" } renders 1`] = `
<div>
  <div
    class="css-hckpid"
  >
    <a
      class="css-zdzv7c"
      href="/"
    >
      Read in English
    </a>
  </div>
</div>
`;
